### PR TITLE
Feature/user supplied metrics

### PR DESCRIPTION
**Pull request by improbable-danny**
#### Description
Adding support for user supplied metrics

#### Release note
- Support for metrics to be submitted by game code. Note this is only available for already approved Improbable metrics.

#### Tests
Tested using NFR gym

STRONGLY SUGGESTED: How can this be verified by QA?

It can be verified by submitting a metric approved by Prometheus 

#### Primary reviewers
@m-samiec

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -31,6 +31,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - You can now specify deployment tags in the `Cloud Deployment` window.
 - RPCs declared in a UINTERFACE can now be executed. Previously, this would lead to a runtime assertion.
 - When using the `-receptionistHost` command line parameter with a non-local host, it's no longer necessary to set `-useExternalIpForBridge true` as this will be inferred automatically.
+- Support for metrics to be submitted by game code. Note this is only available for already approved Improbable metrics.
```
Comment (improbable-danny) [inline CHANGELOG.md:34]: Should we publicize this?

Comment (m-samiec) [reply CHANGELOG.md:34]: Yeah I don't see why not. Should specify what the acceptable metrics are, and potentially create vars for them in `USpatialMetrics`

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp [hunk 2]
```diff
@@ -51,18 +51,41 @@ void USpatialMetrics::TickMetrics(float NetDriverTime)
...
+	// Dynamic.FPS
+	{
+		SpatialGDK::GaugeMetric DynamicFPSGauge;
+		DynamicFPSGauge.Key = TCHAR_TO_UTF8(*SpatialConstants::SPATIALOS_METRICS_DYNAMIC_FPS);
+		DynamicFPSGauge.Value = AverageFPS;
+		Metrics.GaugeMetrics.Add(DynamicFPSGauge);
+	}
+	// User supplied metrics
+	TArray<FString> UnboundMetrics;
+	for (TPair<FString, UserSppliedMetric> Guage : UserSuppliedMetrics)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp:65]:
```suggestion
	for (const TPair<FString, UserSppliedMetric>& Gauge : UserSuppliedMetrics)
```

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialMetrics.h [hunk 3]
```diff
@@ -57,7 +57,9 @@ class SPATIALGDK_API USpatialMetrics : public UObject
 	DECLARE_DELEGATE_RetVal(FUnrealObjectRef, FControllerRefProviderDelegate);
 	FControllerRefProviderDelegate ControllerRefProvider;
 
-	void SetWorkerLoadDelegate(const WorkerLoadFunction& Delegate) { WorkerLoadDelegate = Delegate; }
+	void SetWorkerLoadDelegate(const UserSppliedMetric& Delegate) { WorkerLoadDelegate = Delegate; }
+	void AddCustomMetric(const TCHAR* Metric, const UserSppliedMetric& Delegate);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialMetrics.h:61]: These two can just operate on FStrings

File: SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialMetrics.h [hunk 4]
```diff
@@ -73,7 +75,9 @@ class SPATIALGDK_API USpatialMetrics : public UObject
 
 	double AverageFPS;
 	double WorkerLoad;
-	WorkerLoadFunction WorkerLoadDelegate;
+	UserSppliedMetric WorkerLoadDelegate;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Public/Utils/SpatialMetrics.h:78]:
```suggestion
	UserSuppliedMetric WorkerLoadDelegate;
```

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp [hunk 5]
```diff
@@ -51,18 +55,34 @@ void USpatialMetrics::TickMetrics(float NetDriverTime)
...
 	}
 
-	SpatialGDK::GaugeMetric DynamicFPSGauge;
-	DynamicFPSGauge.Key = TCHAR_TO_UTF8(*SpatialConstants::SPATIALOS_METRICS_DYNAMIC_FPS);
-	DynamicFPSGauge.Value = AverageFPS;
+	SpatialGDK::SpatialMetrics Metrics;
+	Metrics.Load = WorkerLoad; 
+	// User supplied metrics
+	TArray<FString> UnboundMetrics;
+	for (const TPair<FString, UserSuppliedMetric>& Guage : UserSuppliedMetrics)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp:62]:
```suggestion
	for (const TPair<FString, UserSuppliedMetric>& Gauge : UserSuppliedMetrics)
```

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp [hunk 6]
```diff
@@ -51,18 +55,34 @@ void USpatialMetrics::TickMetrics(float NetDriverTime)
 		WorkerLoad = CalculateLoad();
 	}
 
-	SpatialGDK::GaugeMetric DynamicFPSGauge;
-	DynamicFPSGauge.Key = TCHAR_TO_UTF8(*SpatialConstants::SPATIALOS_METRICS_DYNAMIC_FPS);
-	DynamicFPSGauge.Value = AverageFPS;
+	SpatialGDK::SpatialMetrics Metrics;
+	Metrics.Load = WorkerLoad;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp:59]:
```suggestion
	Metrics.Load = WorkerLoad;
	
```

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp [hunk 7]
```diff
@@ -51,18 +55,34 @@ void USpatialMetrics::TickMetrics(float NetDriverTime)
...
+			Metric.Key = TCHAR_TO_UTF8(*Guage.Key);
+			Metric.Value = Guage.Value.Execute();
+			Metrics.GaugeMetrics.Add(Metric);
+		}
+		else
+		{
+			UnboundMetrics.Add(Guage.Key);
+		}
+	}
+	for (FString& KeyToRemove : UnboundMetrics)
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp:77]:
```suggestion
	for (const FString& KeyToRemove : UnboundMetrics)
```

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp [hunk 8]
```diff
@@ -51,18 +55,35 @@ void USpatialMetrics::TickMetrics(float NetDriverTime)
...
+	for (const TPair<FString, UserSuppliedMetric>& Gauge : UserSuppliedMetrics)
+	{
+		if (Guage.Value.IsBound())
+		{
+			SpatialGDK::GaugeMetric Metric;
 
-	SpatialGDK::SpatialMetrics DynamicFPSMetrics;
-	DynamicFPSMetrics.GaugeMetrics.Add(DynamicFPSGauge);
-	DynamicFPSMetrics.Load = WorkerLoad;
+			Metric.Key = TCHAR_TO_UTF8(*Guage.Key);
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp:69]:
```suggestion
			Metric.Key = TCHAR_TO_UTF8(*Guage.Key);
```

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp [hunk 9]
```diff
@@ -51,18 +55,35 @@ void USpatialMetrics::TickMetrics(float NetDriverTime)
...
+			Metric.Value = Guage.Value.Execute();
+			Metrics.GaugeMetrics.Add(Metric);
+		}
+		else
+		{
+			UnboundMetrics.Add(Guage.Key);
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialMetrics.cpp:75]:
```suggestion
			UnboundMetrics.Add(Gauge.Key);
```

**Pull request merged**